rbm_softmax.py

```diff
@@ -0,0 +1,504 @@
+        validation : {array-like, sparse matrix}
+
+        Returns
+        -------
+        self : BernoulliRBM
+            The fitted model.
+        """
+        X = check_array(X, accept_sparse='csr', dtype=np.float)
+        n_samples = X.shape[0]
+
+        if not hasattr(self, 'components_'):
+            self.components_ = np.asarray(
+                self.rng_.normal(0, 0.01, (self.n_components, X.shape[1])),
+                order='fortran')
+            self.intercept_hidden_ = np.zeros(self.n_components, )
+            #self.intercept_visible_ = np.zeros(X.shape[1], )
+            # 'It is usually helpful to initialize the bias of visible unit i to log[p_i/(1-p_i)] where p_i is the prptn of training vectors where i is on' - Practical Guide
+            counts = X.sum(axis=0).A.reshape(-1)
+            # There should be no units that are always on
+            assert np.max(counts) < X.shape[0], "Found a visible unit always on in the training data. Fishy."
+            # There might be some units never on. Add a pseudo-count of 1 to avoid inf
+            vis_priors = (counts + 1) / float(X.shape[0])
+            self.intercept_visible_ = np.log( vis_priors / (1 - vis_priors) )
+
+
+        else:
+            print "Resuing existing weights and biases"
+        # Don't necessarily want to reuse h_samples if we have one leftover from before - batch size might have changed
+        self.h_samples_ = np.zeros((self.batch_size, self.n_components))
+
+        n_batches = int(np.ceil(float(n_samples) / self.batch_size))
+        batch_slices = list(gen_even_slices(n_batches * self.batch_size,
+                                            n_batches, n_samples))
+        verbose = self.verbose
+        begin = time.time()
+        for iteration in xrange(1, self.n_iter + 1):
+            if self.lr_backoff:
+                # If, e.g., we're doing 10 epochs, use the full learning rate for
+                # the first iteration, 90% of the base learning rate for the second
+                # iteration... and 10% for the final iteration
+                self.learning_rate = ((self.n_iter - (iteration - 1)) / (self.n_iter+0.0)) * self.base_learning_rate
+                print "Using learning rate of {:.3f} (base LR={:.3f})".format(self.learning_rate, self.base_learning_rate)
+
+            for batch_slice in batch_slices:
+                self._fit(X[batch_slice])
+
+            if verbose:
+                end = time.time()
+                self.wellness_check(iteration, end - begin, X, validation)
+                begin = end
+
+        return self
+
+    def wellness_check(self, epoch, duration, train, validation):
+        """Log some diagnostic information on how the model is doing so far."""
+        validation_debug = ''
+        if validation is not None:
+            v_energy, t_energy = self.score_validation_data(train, validation)
+            # TODO: Not clear whether relative or absolute difference is the more relevant metric here.
```

Comment:
@falsifian do you have an intuition about this? I prefer absolute difference, because subsequent epochs tend to increase that, while the relative difference sometimes stays the same or goes down. And I'm an optimist.